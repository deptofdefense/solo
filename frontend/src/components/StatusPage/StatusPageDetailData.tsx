import React from "react";
import { formatDistanceToNow, parseISO } from "date-fns";
import { Address, Status, Part } from "solo-types";

interface StatusPageDetailDataProps {
  address: Address[];
  status: Status[];
  part: Part[];
}

const StatusPageDetailData: React.FC<StatusPageDetailDataProps> = ({
  address,
  status,
  part
}) => {
  const shipperAddy = address.find(addy => addy.address_type === "2")?.name;
  const receiverAddy = address.find(addy => addy.address_type === "3")?.name;
  const quantity = status[0]?.qty || 0;

  return (
    <div className="grid-row flex-justify flex-align-start padding-5">
      <div className="flex-col">
        <div className="text-bold">DIC</div>
        {status.map(({ status_date, dic: { code } }) => {
          const formattedDate = formatDistanceToNow(parseISO(status_date));
          return (
            <div
              className="grid-row flex-nowrap flex-col flex-align-center"
              key={code}
            >
              <div className="margin-right-2">{code}</div>
              <div>{`${formattedDate} ago`}</div>
            </div>
          );
        })}
      </div>
      {part.length > 0 && (
        <>
          <div className="flex-col">
            <div className="text-bold">NIIN</div>
            <div>{part[0].nsn}</div>
          </div>
          <div className="flex-col">
            <div className="text-bold">Quantity</div>
            <div>{quantity}</div>
          </div>
          <div className="flex-col">
            <div className="text-bold">Unit of Measure</div>
            <div>{part[0].uom}</div>
          </div>
        </>
      )}
      {shipperAddy && (
        <div className="flex-col">
          <div className="text-bold">Shipped From</div>
          <div>{shipperAddy}</div>
        </div>
      )}
      {receiverAddy && (
        <div className="flex-col">
          <div className="text-bold">Shipped To</div>
          <div>{receiverAddy}</div>
        </div>
      )}
    </div>
  );
};

export default StatusPageDetailData;
