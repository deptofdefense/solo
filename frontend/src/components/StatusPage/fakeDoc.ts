import { Document } from "solo-types";
import { parseISO, sub } from "date-fns";

export const defaultDoc = {
  id: 1,
  part: [
    {
      id: 2,
      nsn: "5930015931045",
      nomen: "Switch, rotary, 24V dc waterproof",
      uom: "ea",
      document: [1]
    }
  ],
  status: [
    {
      id: 1,
      dic: {
        id: 1,
        code: "AS1",
        desc: "Shipping Status",
        status: [1]
      },
      status_date: "2020-03-06T13:01:21-05:00",
      key_and_transmit_date: null,
      esd: null,
      qty: 5,
      document: 1,
      user: null
    },
    {
      id: 2,
      dic: {
        id: 1,
        code: "AS2",
        desc: "Shipping Status",
        status: [1]
      },
      status_date: "2020-03-06T13:01:21-05:00",
      key_and_transmit_date: null,
      esd: null,
      qty: 5,
      document: 1,
      user: null
    }
  ],
  address: [
    {
      id: 7,
      address_type: "1",
      name: "requestor",
      ric: "SMS",
      addy1: "addy1",
      addy2: "addy2",
      addy3: "addy3",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    },
    {
      id: 8,
      address_type: "2",
      name: "AAC-Residence Inn Arlington Pentagon City: Rm 1103",
      ric: "5E3",
      addy1: "Residence Inn Arlington Pentagon City: Rm 1103",
      addy2: "550 Army Navy Dr",
      addy3: "addy3",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    },
    {
      id: 9,
      address_type: "3",
      name: "AAC-M30300",
      ric: "SMS",
      addy1: "addy1",
      addy2: "addy2",
      addy3: "addy3",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    },
    {
      id: 10,
      address_type: "4",
      name: "AAC-M30300",
      ric: "5E3",
      addy1: "addy1",
      addy2: "addy2",
      addy3: "bldg 24006 Montezuma",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    }
  ],
  sdn: "M0000000000000",
  service_request: 5
};

const createFakeDocuments = (count: number = 20): Document[] => {
  return Array.from({ length: count }).map(() => ({
    ...defaultDoc,
    status: defaultDoc.status.map(stat => ({
      ...stat,
      status_date: sub(parseISO(stat.status_date), {
        days: Math.floor(Math.random() * 10)
      }).toISOString()
    }))
  }));
};

export default createFakeDocuments;
