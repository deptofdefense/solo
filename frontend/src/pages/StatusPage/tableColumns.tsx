import React from "react";
import { formatDistanceToNow, parseISO } from "date-fns";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faPlus, faMinus } from "@fortawesome/free-solid-svg-icons";
import { Column } from "react-table";
import { Document } from "solo-types";

type CreateColumns = () => Column<Document>[];

const createColumns: CreateColumns = () => [
  {
    Header: "Details",
    Cell: ({ row }) => (
      <span {...row.getToggleRowExpandedProps()}>
        <FontAwesomeIcon icon={row.isExpanded ? faMinus : faPlus} />
      </span>
    )
  },
  {
    Header: "SDN",
    accessor: "sdn"
  },
  {
    Header: "Service Request #",
    accessor: "serviceRequest.service_request",
    id: "service_request__service_request"
  },
  {
    Header: "Commodity",
    id: "suppadd__code",
    accessor: "commodityName"
  },
  {
    Header: "Status",
    disableSortBy: true,
    id: "currentStatus",
    accessor: ({ mostRecentStatusIdx, statuses }) =>
      statuses[mostRecentStatusIdx].dic.desc
  },
  {
    Header: "Nomenclature",
    id: "part__nomen",
    accessor: "part.nomen"
  },
  {
    Header: "Last Updated",
    id: "statuses__status_date",
    disableSortBy: true,
    accessor: ({ mostRecentStatusIdx, statuses }) =>
      `${formatDistanceToNow(
        parseISO(statuses[mostRecentStatusIdx].status_date)
      )} ago`
  }
];

export default createColumns;
