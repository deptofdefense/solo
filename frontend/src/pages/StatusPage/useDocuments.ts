import { useCallback, useEffect, useState } from "react";
import { SortingRule } from "react-table";
import useAuthContext from "context/AuthContext";
import { Document, ApiDocument } from "solo-types";
import { createFakeApiDocs } from "solo-types";

// covert api returned document to Document
export const parseApiDocuments = (apiDocs: ApiDocument[]): Document[] =>
  apiDocs.map(apiDoc => ({
    id: apiDoc.id,
    sdn: apiDoc.sdn,
    serviceRequest: apiDoc.service_request,
    part: apiDoc.part,
    statuses: apiDoc.statuses,
    suppadd: {
      ...apiDoc.suppadd
    },
    shipper: apiDoc.addresses.find(
      addy => addy.address_type.type === "Ship-To"
    ),
    reciever: apiDoc.addresses.find(
      addy => addy.address_type.type === "Requester"
    )
  }));

const useDocuments = () => {
  const { apiCall } = useAuthContext();
  const [docs, setDocs] = useState<Document[]>([]);
  const [sortBy, setSortBy] = useState<SortingRule<Document>[]>([]);

  const makeQueryString = useCallback(() => {
    const params = new URLSearchParams();
    const currentSort = sortBy[0];
    if (currentSort?.id) {
      params.set("sort", currentSort.id);
    }
    if (currentSort?.desc) {
      params.set("desc", "true");
    }
    const query = params.toString();
    return query ? `?${query}` : "";
  }, [sortBy]);

  const fetchDocuments = useCallback(async () => {
    try {
      const query = makeQueryString();
      const url = `/documents${query}`;
      const docs = await apiCall<ApiDocument[]>(url, {
        method: "GET"
      });
      setDocs(parseApiDocuments(docs));
    } catch (e) {
      // use fake data until api is implemented
      setDocs(parseApiDocuments(createFakeApiDocs(10)));
    }
  }, [setDocs, apiCall, makeQueryString]);

  useEffect(() => {
    fetchDocuments();
  }, [setSortBy, sortBy, fetchDocuments]);

  return {
    docs,
    onSort: setSortBy,
    sort: sortBy
  };
};

export default useDocuments;
