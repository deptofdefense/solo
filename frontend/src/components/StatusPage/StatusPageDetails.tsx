import React from "react";
import { Document } from "solo-types";
import StatusPageDetailData from "./StatusPageDetailData";

interface StatusPageDetailsProps {
  document: Document;
}

const StatusPageDetails: React.FC<StatusPageDetailsProps> = ({ document }) => {
  const { status, part, address } = document;

  return (
    <>
      <StatusPageDetailData address={address} part={part} status={status} />
    </>
  );
};

export default StatusPageDetails;
