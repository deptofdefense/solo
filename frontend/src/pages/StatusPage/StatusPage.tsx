import React, { useMemo } from "react";
import { Row, TableInstance } from "react-table";
import { Document } from "solo-types";
import {
  Table,
  DocumentDetails,
  DocumentStepper,
  Title,
  SelectFilterControls,
  Paginator
} from "components";
import useDocuments from "./useDocuments";
import createColumns from "./tableColumns";

const filterAble = [
  { name: "SDN", value: "sdn" },
  { name: "Nomenclature", value: "nomen" },
  { name: "Commodity", value: "commodity" }
];

const StatusPage: React.FC = () => {
  const { docs, updateDocuments, pageCount } = useDocuments();
  const tableColumns = useMemo(createColumns, []);

  const renderSubComponent = (row: Row<Document>) => {
    const {
      original: { shipper, receiver, part, statuses }
    } = row;
    return (
      <>
        <DocumentStepper statuses={statuses} />
        <DocumentDetails
          shipper={shipper}
          receiver={receiver}
          part={part}
          statuses={statuses}
        />
      </>
    );
  };

  const renderPagination = (table: TableInstance<Document>) => (
    <Paginator table={table} />
  );
  const renderFilterControls = (table: TableInstance<Document>) => {
    const { setGlobalFilter } = table;
    return (
      <SelectFilterControls options={filterAble} onSubmit={setGlobalFilter} />
    );
  };

  return (
    <div className="tablet:margin-x-8 overflow-x-auto">
      <Title>Status</Title>
      <Table<Document>
        columns={tableColumns}
        data={docs}
        renderSubComponent={renderSubComponent}
        renderPagination={renderPagination}
        renderFilterControls={renderFilterControls}
        fetchData={updateDocuments}
        pageCount={pageCount}
      />
    </div>
  );
};

export default StatusPage;
