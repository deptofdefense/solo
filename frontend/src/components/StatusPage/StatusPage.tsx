import React, { useMemo, useCallback, useEffect, useState } from "react";
import { Column, Row, SortingRule } from "react-table";
import { formatDistanceToNow, parseISO } from "date-fns";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faPlus, faMinus } from "@fortawesome/free-solid-svg-icons";
import { Document } from "solo-types";
import { Table } from "components";
import useAuthContext from "context/AuthContext";
import createFakeDocs from "./fakeDoc";
import StatusPageDetails from "./StatusPageDetails";

const StatusPage: React.FC = () => {
  const { apiCall } = useAuthContext();
  const [docs, setDocs] = useState<Document[]>([]);
  const [sortBy, setSortBy] = useState<SortingRule<Document>[]>([]);

  const fetchDocuments = useCallback(async () => {
    try {
      const params = new URLSearchParams();
      const sort = sortBy[0];
      if (sort?.id) {
        params.set("sort", sort?.id);
      }
      if (sort?.desc) {
        params.set("desc", "true");
      }
      const query = params.toString();
      const queryString = query ? `?${query}` : "";
      const url = `/documents${queryString}`;
      setDocs(
        await apiCall<Document[]>(url, {
          method: "GET"
        })
      );
    } catch (e) {
      // use fake data until api is implemented
      setDocs(createFakeDocs(10));
    }
  }, [setDocs, sortBy, apiCall]);

  useEffect(() => {
    fetchDocuments();
  }, [setSortBy, sortBy, fetchDocuments]);

  const columns: Column<Document>[] = useMemo(
    () => [
      {
        Header: "Details",
        Cell: ({ row }) => (
          <span {...row.getToggleRowExpandedProps()}>
            <FontAwesomeIcon icon={row.isExpanded ? faMinus : faPlus} />
          </span>
        )
      },
      {
        Header: "SDN",
        accessor: "sdn"
      },
      {
        Header: "Service Request #",
        accessor: "service_request"
      },
      {
        Header: "Commodity",
        id: "commodity",
        accessor: () => "Motor T"
      },
      {
        Header: "Status",
        disableSortBy: true,
        accessor: "status[0].dic[0].desc"
      },
      {
        Header: "Nomenclature",
        id: "nomen",
        accessor: "part[0].nomen"
      },
      {
        Header: "Last Updated",
        id: "status_date",
        accessor: ({ status }) =>
          `${formatDistanceToNow(parseISO(status[0].status_date))} ago`
      }
    ],
    []
  );

  return (
    <div className="tablet:margin-x-8 overflow-x-auto">
      <Table<Document>
        columns={columns}
        data={docs}
        renderSubComponent={(row: Row<Document>) => (
          <StatusPageDetails document={row.original} />
        )}
        onSort={setSortBy}
        initialSortBy={sortBy}
      />
    </div>
  );
};

export default StatusPage;
