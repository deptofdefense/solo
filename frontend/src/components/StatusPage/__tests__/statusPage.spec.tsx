import React from "react";
import { render } from "test-utils";
import StatusPage from "../StatusPage";

describe("StatusPage component", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(<StatusPage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
