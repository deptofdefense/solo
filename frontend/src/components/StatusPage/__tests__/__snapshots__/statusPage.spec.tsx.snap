// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusPage component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    Status Page
  </div>
</DocumentFragment>
`;
